services/operator_info.go

```diff
@@ -0,0 +1,134 @@
+package services
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/Rocket-Rescue-Node/credentials"
+	"github.com/Rocket-Rescue-Node/credentials/pb"
+	"github.com/Rocket-Rescue-Node/rescue-api/models"
+	authz "github.com/Rocket-Rescue-Node/rescue-api/models/authorization"
+	"github.com/Rocket-Rescue-Node/rescue-api/util"
+	"go.uber.org/zap"
+)
+
+const (
+	// The maximum age for a operator info request to be considered valid.
+	operatorInfoRequestMaxAge = time.Duration(15) * time.Minute
+)
+
+type OperatorInfo struct {
+	CredentialEvents []int64 `json:"credentialEvents"`
+	QuotaSettings    *Quota  `json:"quotaSettings,omitempty"`
+}
+
+func CreateOperatorInfo(credentialEvents []int64, quotaSettings Quota) (*OperatorInfo, error) {
+	message := OperatorInfo{}
+	message.CredentialEvents = credentialEvents
+	message.QuotaSettings = &quotaSettings
+
+	return &message, nil
+}
+
+func (s *Service) GetOperatorInfo(msg []byte, sig []byte, ot credentials.OperatorType) (*OperatorInfo, error) {
+	var err error
+
+	// Get node address
+	nodeID, err := util.RecoverAddressFromSignature(msg, sig)
+	if err != nil {
+		msg := "failed to recover nodeID from signature"
+		s.logger.Warn(msg, zap.Error(err))
+		s.m.Counter("get_operator_info_failed_auth").Inc()
+		return nil, &AuthenticationError{msg}
+	}
+	s.logger.Info("Recovered nodeID from signature", zap.String("nodeID", nodeID.Hex()))
+
+	// Check if the request is fresh
+	tsSecs, err := s.getTimestampFromRequest(string(msg))
+	if err != nil {
+		s.m.Counter("get_operator_info_invalid_timestamp").Inc()
+		return nil, &ValidationError{"invalid timestamp"}
+	}
+	ts := time.Unix(tsSecs, 0)
+	if time.Since(ts) > operatorInfoRequestMaxAge {
+		s.m.Counter("get_operator_info_timestamp_too_old").Inc()
+		return nil, &AuthenticationError{"timestamp is too old"}
+	}
+
+	// Check if this node is part of Rocket Pool, or a valid 0x01 credential.
+	switch ot {
+	case pb.OperatorType_OT_ROCKETPOOL:
+		if !s.isNodeRegistered(nodeID) {
+			s.m.Counter("get_operator_info_node_not_registered").Inc()
+			return nil, &AuthorizationError{"node is not registered"}
+		}
+	case pb.OperatorType_OT_SOLO:
+		if !s.enableSoloValidators {
+			s.m.Counter("get_operator_info_solo_traffic_shedding").Inc()
+			return nil, &AuthorizationError{"solo validators are currently not permitted"}
+		}
+		if !s.isWithdrawalAddress(nodeID) {
+			s.m.Counter("get_operator_info_solo_not_withdrawal_address").Inc()
+			return nil, &AuthorizationError{"wallet is not a withdrawal address for any validator"}
+		}
+	}
```

Comment:
I don't care enough about granular metrics to not say "Hey, let's make a function that deduplicates this logic from here and its counterpart in the creds service"

Really, deduplicate as much of this logic as you can across the two services. Each should have a function that calls some precondition validation and the logic only differs at the time we query the database.

Reply:
Took a crack at this. Could have abstracted less / slightly more, but I felt my attempt is easy to understand for future maintenance.